File: RepoTransBench/target_projects/Python/Go/nateshmbhat_pyttsx3/public_tests/public_driverproxy_api_test.go
package public_tests

import (
	"reflect"
	"testing"
)

// --- Types as in public python tests ---

type AnotherDummyEngine struct {
	Notified []Notification
}
type Notification struct {
	Topic string
	Kw    map[string]interface{}
}
func (e *AnotherDummyEngine) Notify(topic string, kw map[string]interface{}) {
	e.Notified = append(e.Notified, Notification{Topic: topic, Kw: kw})
}

type AnotherDummyDriver struct {
	Proxy    *DriverProxy
	Said     [][2]string
	Stopped  bool
	Busy     *bool
	SaidItems [][2]string
	TimesStopped int
}
func (d *AnotherDummyDriver) StartLoop()            {}
func (d *AnotherDummyDriver) EndLoop()              {}
func (d *AnotherDummyDriver) Say(text, name string) { d.Said = append(d.Said, [2]string{text, name}) }
func (d *AnotherDummyDriver) Stop()                 { d.Stopped = true; d.TimesStopped += 1 }
func (d *AnotherDummyDriver) SetBusy(b bool)        { d.Busy = &b }
func (d *AnotherDummyDriver) NotifyIt(data interface{}) {}

// ---- DriverProxy ----

type QueuedFunc struct {
	Fn   func(...interface{}) interface{}
	Args []interface{}
	Name string
}
type DriverProxy struct {
	Engine *AnotherDummyEngine
	Driver *AnotherDummyDriver
	Busy   bool
	Queue  []QueuedFunc
	Name   string
}
func NewDriverProxyPublic(engine *AnotherDummyEngine, name string) *DriverProxy {
	return &DriverProxy{
		Engine: engine,
		Driver: &AnotherDummyDriver{},
		Busy:   true,
		Queue:  make([]QueuedFunc, 0),
		Name:   name,
	}
}
func (p *DriverProxy) Push(fn func(...interface{}) interface{}, args []interface{}, name string) {
	p.Queue = append(p.Queue, QueuedFunc{Fn: fn, Args: args, Name: name})
}
func (p *DriverProxy) Notify(topic string, kw map[string]interface{}) {
	if kw == nil {
		kw = map[string]interface{}{}
	}
	kw["name"] = p.Name
	p.Engine.Notify(topic, kw)
}
func (p *DriverProxy) SetBusy(b bool) { p.Busy = b }
func (p *DriverProxy) IsBusy() bool   { return p.Busy }
func (p *DriverProxy) Say(text, name string) {
	// Emulate queued call
	p.Queue = append(p.Queue, QueuedFunc{
		Fn: func(args ...interface{}) interface{} {
			p.Driver.Say(text, name)
			return nil
		}, Args: []interface{}{text, name}, Name: name,
	})
}
func (p *DriverProxy) Stop() {
	p.Queue = append(p.Queue, QueuedFunc{
		Fn: func(args ...interface{}) interface{} {
			p.Driver.Stop()
			return nil
		}, Args: []interface{}{}, Name: "",
	})
}
func (p *DriverProxy) Del() {
	p.Driver.Stopped = true // as placeholder for del/destroy
}

func TestPublicDriverproxyInit(t *testing.T) {
	eng := &AnotherDummyEngine{}
	proxy := NewDriverProxyPublic(eng, "otherdummy")
	if reflect.TypeOf(proxy.Driver).Elem().Name() != "AnotherDummyDriver" {
		t.Errorf("Wrong driver type on init")
	}
	if proxy.Engine != eng {
		t.Errorf("Engine not set")
	}
	if !proxy.Busy {
		t.Errorf("Should be busy by default")
	}
}

func TestPublicDriverproxyDel(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "del")
	defer func() {
		if v := recover(); v != nil {
			t.Errorf("del raised panic: %v", v)
		}
	}()
	proxy.Del() // Should not panic
	if !proxy.Driver.Stopped {
		t.Errorf("Driver.Del did not set stopped to true")
	}
}

func TestPublicDriverproxyPushAndPump(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "push")
	proxy.Push(func(args ...interface{}) interface{} {
		return args[0].(string) + "X"
	}, []interface{}{"fox"}, "")
	proxy.Push(func(args ...interface{}) interface{} {
		// Reverse string
		s := args[0].(string)
		runes := []rune(s)
		for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
			runes[i], runes[j] = runes[j], runes[i]
		}
		return string(runes)
	}, []interface{}{"bottle"}, "")
	collected := []interface{}{}
	for len(proxy.Queue) > 0 {
		qf := proxy.Queue[0]
		proxy.Queue = proxy.Queue[1:]
		collected = append(collected, qf.Fn(qf.Args...))
	}
	if collected[0] != "foxX" {
		t.Errorf("Expected foxX, got %v", collected[0])
	}
	if collected[1] != "elttob" {
		t.Errorf("Expected elttob, got %v", collected[1])
	}
}

func TestPublicDriverproxyNotify(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "notifyName")
	proxy.Notify("pub_new_notify", map[string]interface{}{"key": "val"})
	if len(proxy.Engine.Notified) == 0 {
		t.Fatalf("Engine not notified")
	}
	last := proxy.Engine.Notified[len(proxy.Engine.Notified)-1]
	if last.Topic != "pub_new_notify" {
		t.Errorf("Expected topic pub_new_notify, got %v", last.Topic)
	}
	if last.Kw["key"] != "val" {
		t.Errorf("Expected key=val, got %v", last.Kw["key"])
	}
}

func TestPublicDriverproxySetbusyAndIsbusy(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "busy")
	proxy.SetBusy(false)
	if proxy.IsBusy() {
		t.Errorf("isBusy should be false after setBusy(false)")
	}
	proxy.SetBusy(true)
	if !proxy.IsBusy() {
		t.Errorf("isBusy should be true after setBusy(true)")
	}
}

func TestPublicDriverproxySay(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "say")
	proxy.Say("Hi from public!", "pTestName")
	found := false
	for _, q := range proxy.Queue {
		if q.Name == "pTestName" && len(q.Args) > 0 && q.Args[0] == "Hi from public!" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("Say not queued properly")
	}
}

func TestPublicDriverproxyStop(t *testing.T) {
	proxy := NewDriverProxyPublic(&AnotherDummyEngine{}, "stop")
	proxy.Driver.TimesStopped = 0
	proxy.Stop()
	for len(proxy.Queue) > 0 {
		qf := proxy.Queue[0]
		proxy.Queue = proxy.Queue[1:]
		if qf.Fn != nil {
			qf.Fn()
		}
	}
	if proxy.Driver.TimesStopped != 1 {
		t.Errorf("Expected driver times_stopped=1, got %d", proxy.Driver.TimesStopped)
	}
}

File: RepoTransBench/target_projects/Python/Java/nateshmbhat_pyttsx3/src/test/java/com/nateshmbhat/pyttsx3/publictests/PublicDriverProxyApiTest.java
package com.nateshmbhat.pyttsx3.publictests;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;

import java.util.*;
import java.util.Map.Entry;

class AnotherDummyEngineProxy {
    List<Entry<String, Map<String, Object>>> notified = new ArrayList<>();
    public void _notify(String topic, Map<String, Object> kwargs) {
        notified.add(new AbstractMap.SimpleEntry<>(topic, kwargs));
    }
}

class AnotherDummyDriverProxy {
    Object proxy;
    List<String[]> said = new ArrayList<>();
    boolean stopped = false;
    Boolean busy = null;
    int timesStopped = 0;

    public AnotherDummyDriverProxy(Object proxy) { this.proxy = proxy; }
    public void startLoop() {}
    public void endLoop() {}
    public void say(String text, String name) {
        said.add(new String[]{text, name});
    }
    public void stop() {
        stopped = true;
        timesStopped += 1;
    }
    public void setBusy(Boolean value) { busy = value; }
    public void notify(Object data) {}
}

interface PublicTestMethod {
    Object call(Object[] args);
}

class PublicDriverProxy {
    final AnotherDummyEngineProxy _engine;
    final String _name;
    AnotherDummyDriverProxy _driver;
    boolean _busy = true;
    List<Object[]> _queue = new ArrayList<>();

    public PublicDriverProxy(AnotherDummyEngineProxy eng, String name, boolean debug, Class<? extends AnotherDummyDriverProxy> driverCls) {
        this._engine = eng;
        this._name = name;
        try {
            this._driver = driverCls.getConstructor(Object.class).newInstance(this);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        this._busy = true;
        this._queue = new ArrayList<>();
    }

    public void _push(PublicTestMethod meth, Object[] args, String name) {
        _queue.add(new Object[]{meth, args, name});
    }

    public void notify(String topic, Map<String, Object> kw) {
        kw.put("name", this._name);
        this._engine._notify(topic, kw);
    }

    public void setBusy(boolean val) { this._busy = val; }
    public boolean isBusy() { return this._busy; }

    public void say(String text, String tid) {
        this._driver.say(text, tid);
        this._push((arguments) -> { _driver.say((String)arguments[0], (String)arguments[1]); return null; }, new Object[]{text, tid}, tid);
    }

    public void stop() {
        this._driver.stop();
        this._push((arguments) -> { _driver.stop(); return null; }, new Object[]{}, "stop");
    }
    public void __del__() {}
}

public class PublicDriverProxyApiTest {

    PublicDriverProxy baseDriverProxyForMagic(String testName, Class<? extends AnotherDummyDriverProxy> driverCls) {
        return new PublicDriverProxy(new AnotherDummyEngineProxy(), testName, true, driverCls);
    }

    @Test
    void testPublicDriverProxyInit() {
        PublicDriverProxy proxy = new PublicDriverProxy(new AnotherDummyEngineProxy(), "otherdummy", true, AnotherDummyDriverProxy.class);
        assertTrue(proxy._driver instanceof AnotherDummyDriverProxy);
        assertSame(proxy._engine, proxy._engine);
        assertTrue(proxy._busy);
    }

    @Test
    void testPublicDriverProxyDel() {
        PublicDriverProxy proxy = baseDriverProxyForMagic("del", AnotherDummyDriverProxy.class);
        proxy.__del__(); // Should not throw
    }

    @Test
    void testPublicDriverProxyPushAndPump() {
        PublicDriverProxy proxy = baseDriverProxyForMagic("push", AnotherDummyDriverProxy.class);
        proxy._push((args) -> ((String)args[0]).toUpperCase(), new Object[]{"fox"}, "q1");
        proxy._push((args) -> new StringBuilder((String)args[0]).reverse().toString(), new Object[]{"bottle"}, "q2");
        List<Object> collected = new ArrayList<>();
        while (!proxy._queue.isEmpty()) {
            Object[] q = proxy._queue.remove(0);
            PublicTestMethod func = (PublicTestMethod) q[0];
            Object[] args = (Object[])q[1];
            collected.add(func.call(args));
        }
        assertEquals("FOX", collected.get(0));
        assertEquals("elttob", collected.get(1));
    }

    @Test
    void testPublicDriverProxyNotify() {
        PublicDriverProxy proxy = baseDriverProxyForMagic("notify", AnotherDummyDriverProxy.class);
        Map<String, Object> kw = new HashMap<>();
        kw.put("key", "val");
        proxy.notify("pub_new_notify", kw);
        Entry<String, Map<String, Object>> notific = proxy._engine.notified.get(proxy._engine.notified.size() - 1);
        assertEquals("pub_new_notify", notific.getKey());
        assertEquals("val", notific.getValue().get("key"));
    }

    @Test
    void testPublicDriverProxySetBusyAndIsBusy() {
        PublicDriverProxy proxy = baseDriverProxyForMagic("busy", AnotherDummyDriverProxy.class);
        proxy.setBusy(false);
        assertFalse(proxy.isBusy());
        proxy.setBusy(true);
        assertTrue(proxy.isBusy());
    }

    @Test
    void testPublicDriverProxySay() {
        class SayDriver extends AnotherDummyDriverProxy {
            List<String[]> saidItems = new ArrayList<>();
            public SayDriver(Object proxy) { super(proxy); }
            @Override
            public void say(String text, String name) {
                saidItems.add(new String[]{text, name});
            }
        }
        PublicDriverProxy proxy = baseDriverProxyForMagic("say", SayDriver.class);
        proxy.say("Hi from public!", "pTestName");
        boolean found = false;
        for (Object[] q : proxy._queue) {
            PublicTestMethod func = (PublicTestMethod)q[0];
            Object[] args = (Object[])q[1];
            String text = (String) args[0];
            String name = (String) args[1];
            if (text.equals("Hi from public!") && name.equals("pTestName")) {
                found = true;
            }
        }
        assertTrue(found);
    }

    @Test
    void testPublicDriverProxyStop() {
        class StopDriver extends AnotherDummyDriverProxy {
            int timesStopped = 0;
            public StopDriver(Object proxy) { super(proxy); }
            @Override
            public void stop() {
                timesStopped += 1;
            }
        }
        PublicDriverProxy proxy = baseDriverProxyForMagic("stop", StopDriver.class);
        ((StopDriver)proxy._driver).timesStopped = 0;
        proxy.stop();
        while (!proxy._queue.isEmpty()) {
            Object[] q = proxy._queue.remove(0);
            PublicTestMethod func = (PublicTestMethod)q[0];
            func.call(new Object[]{});
        }
        assertEquals(1, ((StopDriver)proxy._driver).timesStopped);
    }
}